miner.c

```diff
@@ -10508,8 +10505,11 @@ enum test_nonce2_result _test_nonce2(struct work *work, uint32_t nonce, bool che
 			{
 				applog(LOG_DEBUG, "Stratum pool %u target has changed since work job issued, checking that too",
 				       pool->pool_no);
-				if (hash_target_check_v(work->hash, pool->next_target))
+				if (hash_target_check_v(work->hash, pool->next_target)) {
 					high_hash = false;
+					memcpy(work->target, pool->next_target, sizeof(work->target));
+					calc_diff(work, 0);
```

Comment:
I am not completely sure that there are no side effects, but I am pretty sure that when this happens and later those shares are discarded as stale, diff_stale is calculated absolutely incorrectly.
Also, the pool does not need to be broken. The code basically says: shares are good to be sent to pool if they comply with work->work_difficulty OR pool->next_target.
Let us consider the following scenario:
  - nonce_diff is 1. pool starts with difficulty 8. That difficulty is written down as work->work_difficulty. Then pool changes difficulty to 1. pool->next_target = 1. Then at some moment a nonce is reported with diff 1.5. Test_nonce will return TNR_GOOD, because of pool->next_target.
 - pool connection breaks, miner is unable to send that share to the pool. A lot of those shares. All of them are cleared after a timeout. And they are accumulated in stale_diff counter with weight 8, although in reality they were checked against diff 1! Now we have stale_diff 8 times greater then it should be.

That is a scenario I am aiming with that change. I did not see negative side effects in my experiments, but again, I am not completely sure that there are no any.